proposals/0021-root-signature-in-clang.md

```diff
@@ -0,0 +1,314 @@
+into a global variable with struct type to express the layout and metadata to
+save things like static sampler, root flags, space and NumDescriptors in LLVM IR.
+
+CGHLSLRuntime will generate metadata to link the global variable as root
+signature for given entry function.
+
+For case compile to a standalone root signature blob, the
+HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+In the case of local or global root signatures, only the HLSLRootSignatureDecl 
+will be created.
+
+Additionally, a new AST node, SubobjectToExportsAssociationDecl, should be 
+added to handle the SubobjectToExportsAssociation statement in HLSL. 
+The SubobjectToExportsAssociationDecl will contain a HLSLRootSignatureDecl and 
+a list of FunctionDecls.
+
+The default root signature version number is 2 which maps to rootsig_1_1.
+With option -force-rootsig-ver rootsig_1_0 could make it to 1.
+
+## Detailed design
+
+
+* Example root signature
+
+```
+
+    #define RS "RootFlags( ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT | " \
+              "DENY_VERTEX_SHADER_ROOT_ACCESS), " \
+              "CBV(b0, space = 1, flags = DATA_STATIC), " \
+              "SRV(t0), " \
+              "UAV(u0), " \
+              "DescriptorTable( CBV(b1), " \
+              "                 SRV(t1, numDescriptors = 8, " \
+              "                     flags = DESCRIPTORS_VOLATILE), " \
+              "                 UAV(u1, numDescriptors = unbounded, " \
+              "                     flags = DESCRIPTORS_VOLATILE)), " \
+              "DescriptorTable(Sampler(s0, space=1, numDescriptors = 4)), " \
+              "RootConstants(num32BitConstants=3, b10), " \
+              "StaticSampler(s1)," \
+              "StaticSampler(s2, " \
+              "              addressU = TEXTURE_ADDRESS_CLAMP, " \
+              "              filter = FILTER_MIN_MAG_MIP_LINEAR )"
+
+```
+
+* Define of HLSLEntryRootSignatureAttr
+
+```
+
+    def HLSLEntryRootSignature: InheritableAttr {
+      let Spellings = [GNU<"RootSignature">];
+      let Subjects = Subjects<[HLSLEntry]>;
+      let LangOpts = [HLSL];
+      let Args = [StringArgument<"InputString">, DeclArgument<HLSLRootSignature, "RootSignature", 0, /*fake*/ 1>];
+    }
+
+```
+
+* HLSLEntryRootSignatureAttr in AST.
+
+```
+
+    HLSLEntryRootSignatureAttr
+      "RootFlags( ALLOW_INPUT_ASSEMBLER_INPUT_LAYOUT | DENY_VERTEX_SHADER_ROOT_ACCESS),
+       CBV(b0, space = 1, flags = DATA_STATIC), SRV(t0), UAV(u0),
+       DescriptorTable( CBV(b1), SRV(t1, numDescriptors = 8,flags = DESCRIPTORS_VOLATILE),
+       UAV(u1, numDescriptors = unbounded, flags = DESCRIPTORS_VOLATILE)),
+       DescriptorTable(Sampler(s0, space=1, numDescriptors = 4)),
+       RootConstants(num32BitConstants=3, b10),
+       StaticSampler(s1),
+       StaticSampler(s2, addressU = TEXTURE_ADDRESS_CLAMP, filter = FILTER_MIN_MAG_MIP_LINEAR )"
+       HLSLRootSignature 'main.RS'
+
+```
+
+* LLVM IR for the root signature saved in ConstantStruct.
```

Comment:
Are you ready now to make a decision over whether or not the ConstantStruct approach or the metadata approach is preferable?

To my eyes, the metadata approach is simpler, more concise and I can easily imagine further changes we can make to further simplify it or extend it.

Reply:
I was thinking ConsantStruct could save the serialization by just make the global variable with "RTS0" section. But it does not work because cannot have extra global variable when emit dxil. :(
I'll update the doc to use metadata.

Reply:
Updated.